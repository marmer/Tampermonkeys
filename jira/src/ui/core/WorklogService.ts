export interface Worklog {
    author: {
        displayName: string
    }
    timeSpendInMinutes: number
}

export default class WorklogService {
    getSummedWorklogsByUser(): Worklog[] {
        return [{author: {displayName: "me"}, timeSpendInMinutes: 42},
            {author: {displayName: "myself"}, timeSpendInMinutes: 546},
            {author: {displayName: "and I"}, timeSpendInMinutes: 1337}]
    }
}