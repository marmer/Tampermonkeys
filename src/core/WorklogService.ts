import groupBy from "./groupBy";
import IssueSiteInfos from "./IssueSiteInfos";
import JiraTimeService from "./JiraTimeService";
import UserService from "./UserService";

export interface WorklogSumByUser {
    author: {
        displayName: string,
    };
    timeSpentInMinutes: number;
}

export interface Worklog {
    author: {
        displayName: string,
        name: string;
    };
    timeSpentInMinutes: number;
    started: string;
    id: string;
    comment: string;
    issueId: string;
}

export default class WorklogService {
    public static async createWorklog(worklog: {
        timeSpentInMinutes: number;
        started: string;
        comment: string;
        issueKey: string;
    }): Promise<void> {
        const {comment, issueKey, started, timeSpentInMinutes} = worklog;
        const timeSpent = JiraTimeService.minutesToJiraFormat(timeSpentInMinutes);

        const response = await fetch(IssueSiteInfos.getWorklogUrlForIssueKey(issueKey), {
            method: "POST",
            body: JSON.stringify({
                comment,
                timeSpent,
                started,
            }),
        }).catch(() => {
            throw new Error("Communication error: Worklog creation failed");
        });

        if (response.status !== 201) {
            throw new Error("Unexpected status code. Creation of worklog probably not sucecssful");
        }
    }

    public static async deleteWorklog(worklog: Worklog): Promise<void> {
        const response = await fetch(IssueSiteInfos.getWorklogModifyUrlByWorklog(worklog), {method: "DELETE"})
            .catch(() => {
                throw new Error("Communication error: Worklog creation failed");
            });

        if (response.status !== 204) {
            throw new Error("Unexpected status code. Deletion of worklog probably not sucecssful");
        }
    }

    public static async updateWorklog(worklog: Worklog): Promise<void> {
        // TODO: marmer 08.10.2019 implement
        throw new Error("Not implemented yet");
    }
    public static async getSummedWorklogsByUser(): Promise<WorklogSumByUser[]> {
        const issueKey = IssueSiteInfos.getCurrentIssueKey();
        return this.getWorklogByIssueKey(issueKey).then(this.sumUp);
    }

    public static async getWorklogsForCurrentIssueAndUser(): Promise<Worklog[]> {
        const worklogByIssueKey = await WorklogService.getWorklogByIssueKey(IssueSiteInfos.getCurrentIssueKey());
        const currentUserName = await UserService.getCurrentUserName();
        return worklogByIssueKey.filter(worklog => worklog.author.name === currentUserName);
    }

    private static async getWorklogByIssueKey(issueKey: string): Promise<Worklog[]> {
        return fetch(IssueSiteInfos.getWorklogUrlForIssueKey(issueKey), {method: "GET"})
            .then((response) => {
                if (response.status !== 200) {
                    throw new Error("Unexpected response status: " + response.status);
                }
                return response.json();
            })
            .then((responseJson) => responseJson.worklogs.map(WorklogService.toWorklog));
    }

    private static toWorklog(responseWorklog: any): Worklog {
        const {timeSpentSeconds, comment, author, id, started, issueId} = responseWorklog;
        const timeSpentInMinutes = Math.floor(timeSpentSeconds / 60);
        const {displayName, name} = author;

        return {
            author: {
                displayName,
                name,
            },
            timeSpentInMinutes,
            comment,
            id,
            started,
            issueId,
        };
    }

    private static sumUp(worklog: Worklog[]): WorklogSumByUser[] {
        const worklogSums: WorklogSumByUser[] = worklog.map(wl => ({
            author: {
                displayName: wl.author.displayName,
            },
            timeSpentInMinutes: wl.timeSpentInMinutes,
        }));

        return groupBy(worklogSums, (w) => w.author.displayName)
            .map((tupel) => tupel.values.reduce((previousValue, currentValue) => {
                const newValue = {...previousValue};
                newValue.timeSpentInMinutes += currentValue.timeSpentInMinutes;
                return newValue;
            }))
            .sort((a, b) => a.author.displayName < b.author.displayName ? -1 : 1);
    }
}
