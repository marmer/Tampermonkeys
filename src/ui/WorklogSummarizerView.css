td.timeSpentColumn {
    text-align: right;
}