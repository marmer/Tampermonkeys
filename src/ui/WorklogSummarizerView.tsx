import React, {Component} from 'react';
import WorklogService, {Worklog} from "../core/WorklogService";
import jiraFormat from "../core/jiraFormat";

interface WorklogSummarizerViewState {
    loadingState: "LOADING" | "DONE" | "ERROR",
    worklogs: Worklog[]
}


export interface WorklogSummarizerViewProps {

}

export default class WorklogSummarizerView extends Component<WorklogSummarizerViewProps, WorklogSummarizerViewState> {

    constructor(props: Readonly<WorklogSummarizerViewProps>) {
        super(props);
        this.state = {
            loadingState: "LOADING",
            worklogs: []
        };

        this.loadSummedBookings();
    }

    render(): React.ReactElement {
        switch (this.state.loadingState) {
            case "ERROR":
                return <div>Error. You could try to reaload</div>;
            case "LOADING":
                return <div>Loading. Be patient</div>
            case "DONE":
                return <div>
                    <h1>Worklogs summarized per User</h1>
                    {this.state.loadingState === "DONE" && this.state.worklogs.map(worklog =>
                        <div key={worklog.author.displayName}>
                            {worklog.author.displayName} => {jiraFormat(worklog.timeSpentInMinutes)}
                        </div>)}
                </div>
        }

    }

    private loadSummedBookings() {
        this.setState({
            loadingState: "LOADING",
            worklogs: []
        });
        WorklogService.getSummedWorklogsByUser()
            .then(worklogs => this.setState({
                loadingState: "DONE",
                worklogs: worklogs
            }))
            .catch(reason => {
                this.setState({loadingState: "ERROR"})
                return console.error(reason);
            })
    }

}
