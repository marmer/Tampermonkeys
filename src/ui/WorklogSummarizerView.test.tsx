import fn = jest.fn;
import "@testing-library/jest-dom/extend-expect";
import {cleanup, render} from "@testing-library/react";
import React from "react";
import WorklogService, {Worklog} from "../core/WorklogService";
import WorklogSummarizerView from "./WorklogSummarizerView";

describe("WorklogSummarizerView", () => {
    afterEach(() => cleanup());
    it("should render the worklog entries", () => {
        WorklogService.getSummedWorklogsByUser = fn().mockImplementation(() => [
            {
                timeSpentInMinutes: 5,
                author: {
                    displayName: "Billy Big",
                },
            },
            {
                timeSpentInMinutes: 15,
                author: {
                    displayName: "Claudette clever",
                },
            },
            {
                timeSpentInMinutes: 10,
                author: {
                    displayName: "Andy Anderson",
                },
            }] as Worklog[]);

        const worklogSummarizerView = render(<WorklogSummarizerView/>);
        worklogSummarizerView.find;
    });
});
