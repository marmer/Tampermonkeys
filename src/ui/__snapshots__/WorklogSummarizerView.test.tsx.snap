// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorklogSummarizerView should render the worklog entries: onInitialLoading 1`] = `
<div>
  <div>
    Loading. Be patient!
  </div>
</div>
`;

exports[`WorklogSummarizerView should render the worklog entries: whenLoadingIsDone 1`] = `
<div>
  <div>
    <h1>
      Worklogs summarized per User
    </h1>
    <table>
      <thead>
        <tr>
          <th>
            display name
            ^
          </th>
          <th>
            time spent
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="displayNameColumn"
          >
            Billy Big
          </td>
          <td
            class="timeSpentColumn"
          >
            5m
          </td>
        </tr>
        <tr>
          <td
            class="displayNameColumn"
          >
            Claudette clever
          </td>
          <td
            class="timeSpentColumn"
          >
            15m
          </td>
        </tr>
        <tr>
          <td
            class="displayNameColumn"
          >
            Andy Anderson
          </td>
          <td
            class="timeSpentColumn"
          >
            10m
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`WorklogSummarizerView should render the worklog entries: whenSortedByDisplayName 1`] = `
<div>
  <div>
    <h1>
      Worklogs summarized per User
    </h1>
    <table>
      <thead>
        <tr>
          <th>
            display name
            ^
          </th>
          <th>
            time spent
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="displayNameColumn"
          >
            Andy Anderson
          </td>
          <td
            class="timeSpentColumn"
          >
            10m
          </td>
        </tr>
        <tr>
          <td
            class="displayNameColumn"
          >
            Billy Big
          </td>
          <td
            class="timeSpentColumn"
          >
            5m
          </td>
        </tr>
        <tr>
          <td
            class="displayNameColumn"
          >
            Claudette clever
          </td>
          <td
            class="timeSpentColumn"
          >
            15m
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`WorklogSummarizerView should render the worklog entries: whenSortedByTimeSpent 1`] = `
<div>
  <div>
    <h1>
      Worklogs summarized per User
    </h1>
    <table>
      <thead>
        <tr>
          <th>
            display name
          </th>
          <th>
            time spent
            ^
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="displayNameColumn"
          >
            Billy Big
          </td>
          <td
            class="timeSpentColumn"
          >
            5m
          </td>
        </tr>
        <tr>
          <td
            class="displayNameColumn"
          >
            Andy Anderson
          </td>
          <td
            class="timeSpentColumn"
          >
            10m
          </td>
        </tr>
        <tr>
          <td
            class="displayNameColumn"
          >
            Claudette clever
          </td>
          <td
            class="timeSpentColumn"
          >
            15m
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
